### PR TITLE
Add instance capacity gauge

### PR DESCRIPTION
**Pull request by pkuwm**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Implements one part of #539 : InstanceCapacityGauge

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
We need to monitor instance utilization in purpose of understanding what the instance capacity is.

Changes:
1. Change instance monitor to update capacity.
2. Change typo of API updateAttributeInfo in dynamicMetric
3. Combine max usage and instance capacity update into one api in cluster status monitor.
4. Add unit tests.

### Tests

- [x] The following tests are written for this issue:
TestClusterStatusMonitor.testUpdateInstanceCapacity

- [x] The following is the result of the "mvn test" command on the appropriate module:
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestJobQueueCleanUp.testJobQueueAutoCleanUp » ThreadTimeout Method org.testng....
[INFO]
[ERROR] Tests run: 1061, Failures: 1, Errors: 0, Skipped: 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:11 h
[INFO] Finished at: 2020-01-14T18:25:06-08:00
[INFO] ------------------------------------------------------------------------

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 10.123 s - in org.apache.helix.integration.task.TestJobQueueCleanUp
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.906 s
[INFO] Finished at: 2020-01-14T19:08:42-08:00
[INFO] ------------------------------------------------------------------------

### Commits
- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

**Comment by pkuwm** [comment]
Will rebase the PR once #548 is merged.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java [hunk 1]
```diff
@@ -367,22 +367,28 @@ public void increaseMessageReceived(List<Message> messages) {
...
+   * capacity key. Before calling this API, we assume the instance monitors are already registered
+   * in ReadClusterDataStage. If the monitor is not registered, this instance capacity status update
+   * will fail. Keys for maxUsageMap and capacityMap should be the same.
    *
-   * @param maxUsageMap a map of max capacity usage, {instance: maxCapacityUsage}
+   * @param maxUsageMap A map of max capacity usage, {instance: maxCapacityUsage}
+   * @param capacityMap A map of instance capacity, {instance: {capacity key: capacity value}}
    */
-  public void updateInstanceMaxUsage(Map<String, Double> maxUsageMap) {
+  public void updateInstanceCapacityStatus(Map<String, Double> maxUsageMap,
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java]: I combine max usage and instance capacity update into one api, in the purpose of to avoid one more for loop and synchronized lock for instance capacity update. The downside is, we assume keys (instance names) are the same. If you think we should separate the apis, please comment.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 2]
```diff
@@ -103,7 +103,7 @@ protected synchronized boolean doRegister(Collection<DynamicMetric<?, ?>> dynami
    * @param description description of the MBean
    * @param dynamicMetrics the DynamicMetrics
    */
-  private void updateAttributtInfos(Collection<DynamicMetric<?, ?>> dynamicMetrics,
+  public void updateAttributeInfo(Collection<DynamicMetric<?, ?>> dynamicMetrics,
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: This might need to be a separate PR.. Let me know if it is neccessary.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java [hunk 3]
```diff
@@ -199,6 +203,39 @@ protected synchronized double getMaxCapacityUsageGauge() {
...
+
+      // Update MBean's all attributes.
+      List<DynamicMetric<?, ?>> attributeList = Lists.newArrayList(
+          _totalMessagedReceivedCounter,
+          _disabledPartitionsGauge,
+          _enabledStatusGauge,
+          _onlineStatusGauge,
+          _maxCapacityUsageGauge
+      );
+      attributeList.addAll(_capacityMetricsMap.values());
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: Recreating the whole list is expensive, can we have a map to track all metrics? A map would be easier for you to maintain.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 2 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: 1. protected?
2. Thanks for fixing the typo, let use "updateAttributtesInfo"?
3. You will need to add synchronized if it is called somewhere else outside this class. Actually we need to add this to all the methods that are reading or updating _attributeMap and _mBeanInfo.
And please don't lock globally, a separate lock for _attributeMap and _mBeanInfo would be good enough.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java [hunk 4]
```diff
@@ -367,22 +367,28 @@ public void increaseMessageReceived(List<Message> messages) {
...
+      Map<String, Map<String, Integer>> capacityMap) {
     synchronized (_instanceMonitorMap) {
       for (Map.Entry<String, Double> entry : maxUsageMap.entrySet()) {
```
[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java]: Right. I've put comments for this. 2 for loops increase a bit more time. Will change it.

[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java]: If we only see where the api is called here, the two maps' keys are the same. But we can't guarantee future calls would have the same keys. Why I did this is to try to combine 2 updates into one api and reduce one for loop. But if we need 2 for loops, I would like 2 different APIs separately.

**Comment by pkuwm** [comment]
@jiajunwang PR updated. Thanks for reviewing!

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 5]
```diff
@@ -186,7 +186,7 @@ public AttributeList getAttributes(String[] attributes) {
   }
 
   @Override
-  public MBeanInfo getMBeanInfo() {
+  public synchronized MBeanInfo getMBeanInfo() {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: As we discussed, this sync won't work. Let's just lock the attributes list in the direct access method.

File: helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java [hunk 6]
```diff
@@ -242,16 +248,18 @@ private void reportInstanceCapacityMetrics(ClusterStatusMonitor clusterStatusMon
...
         for (AssignableNode node : clusterModel.getAssignableNodes().values()) {
           String instanceName = node.getInstanceName();
           double usage = node.getProjectedHighestUtilization(Collections.emptyMap());
-          maxUsageMap.put(instanceName, usage);
+          clusterStatusMonitor
+              .updateInstanceCapacityStatus(instanceName, usage, node.getMaxCapacity());
         }
-
-        clusterStatusMonitor.updateInstanceMaxUsage(maxUsageMap);
+      } catch (HelixException ex) {
```
[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java]: I actually wanted to make it more specific. Exception to me, is too generic, which I try to avoid.

File: helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java [hunk 7]
```diff
@@ -242,16 +248,18 @@ private void reportInstanceCapacityMetrics(ClusterStatusMonitor clusterStatusMon
         ClusterModel clusterModel = ClusterModelProvider.generateClusterModelFromCurrentState(
             dataProvider, resourceMap, currentStateAssignment);
 
-        Map<String, Double> maxUsageMap = new HashMap<>();
         for (AssignableNode node : clusterModel.getAssignableNodes().values()) {
           String instanceName = node.getInstanceName();
           double usage = node.getProjectedHighestUtilization(Collections.emptyMap());
```
[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java]: Sure.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 5 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: How about locking a separate object for both _MBeanInfo and _attributeMap? Once one of them is accessed, we lock both. In this case, we could make sure MBeanInfo and attributes are bound together. 
```
private final Object lockObject = new Object();

synchronized (lockObject) {
}

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java [hunk 8]
```diff
@@ -67,14 +74,12 @@ public String metricName() {
...
-  // Counters
-  private SimpleDynamicMetric<Long> _totalMessagedReceivedCounter;
-
-  // Gauges
-  private SimpleDynamicMetric<Long> _enabledStatusGauge;
-  private SimpleDynamicMetric<Long> _disabledPartitionsGauge;
-  private SimpleDynamicMetric<Long> _onlineStatusGauge;
-  private SimpleDynamicMetric<Double> _maxCapacityUsageGauge;
+  // Capacity Gauges
+  private final Map<String, SimpleDynamicMetric<Long>> _capacityMetricsMap;
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: I like the design. But it could be confusing why _capacityMetricsMap is not part of the _longMetricsMap.
Can we make it more differentiate?

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 9]
```diff
@@ -158,41 +155,39 @@ public synchronized void unregister() {
   }
 
   @Override
-  public Object getAttribute(String attribute)
-      throws AttributeNotFoundException, MBeanException, ReflectionException {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Please keep these exceptions in the signature. We might want to complete the logic here by throwing the right exceptions moving forward.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 10]
```diff
@@ -158,41 +155,39 @@ public synchronized void unregister() {
...
+      Object value = getAttribute(attributeName);
+      attributeList.add(new Attribute(attributeName, value));
     }
     return attributeList;
   }
 
   @Override
   public MBeanInfo getMBeanInfo() {
-    return _mBeanInfo;
+    synchronized (_lock) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: As we discussed, this sync seems not working. Just remove it?

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 11]
```diff
@@ -158,41 +155,39 @@ public synchronized void unregister() {
...
+      return _mBeanInfo;
+    }
   }
 
   @Override
-  public void setAttribute(Attribute attribute) throws AttributeNotFoundException,
-      InvalidAttributeValueException, MBeanException, ReflectionException {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: I know you want to keep the code clean, but this list will remind us what to do if we implement this method in the future. So I would prefer to keep them.
No strong preference. You might want more reviewers to check it : )

File: helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * under the License.
+ */
+
+/**
+ * Represents type of mbean metric's value.
+ * DOUBLE: metric value type is double.
+ * LONG: metric value type is long.
+ */
+public enum MetricValueType {
+  DOUBLE, LONG
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java]: What's the difference between this type andy the original Double and Long?

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 13]
```diff
@@ -59,6 +55,7 @@
 
   // Attribute name to the DynamicMetric object mapping
   private final Map<String, DynamicMetric> _attributeMap = new HashMap<>();
+  private final Object _lock = new Object();
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Have you tried what we discussed that we don't lock at all? Since the strong consistency is not possible anyway.

File: helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java (hunk 12 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java]: Why I want to have this enum type is, I want to avoid using `InstanceOf` or `typeOf` to check Double or Long types. This enum makes it easier to check Double or Long without impacting the performance like InstanceOf.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 11 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: In my opinion, an implemented method could throw less exceptions, but not more exceptions than the interface. If the implementation doesn't throw any exception, we don't have to put them with the implementation.
If we want to implement this method in the future, we could definitely refer to the interface.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java (hunk 8 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: Currently I don't have a better name for this. Maybe we can add clearer comments for it?

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: Sure, it should work.
Or call _capacityMetricsMap -> _dynamicCapacityMetricsMap to emphasis that it is changeable?

File: helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java (hunk 12 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/monitoring/metrics/MetricValueType.java]: How about:
ENABLED_STATUS_GAUGE("Enabled", Long.class)
And you don't even need that switch-case block, instead just call the default constructor of the class to create the initial value.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java [hunk 14]
```diff
@@ -367,24 +367,25 @@ public void increaseMessageReceived(List<Message> messages) {
...
-      for (Map.Entry<String, Double> entry : maxUsageMap.entrySet()) {
-        InstanceMonitor monitor = _instanceMonitorMap.get(entry.getKey());
-        if (monitor == null) {
-          LOG.warn("Failed to update max usage because instance monitor is not found, instance: {}.",
-              entry.getKey());
-          continue;
-        }
-        monitor.updateMaxCapacityUsage(entry.getValue());
-      }
+  public void updateInstanceCapacityStatus(String instanceName, double maxUsage,
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java]: Will changing the name of a public method break backward-compatibility?

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java [hunk 15]
```diff
@@ -42,23 +43,29 @@
   /**
    * Metric names for instance capacity.
    */
-  public enum InstanceMonitorMetrics {
+  public enum InstanceMonitorMetric {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: Will changing the name of a public enum break backward-compatibility?

Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: No. This class was added in a previous commit for adding the MaxCapacityUsageGauge. It is not yet deployed into production.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java (hunk 14 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java]: No. It is not yet deployed. This api was added one week ago.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java (hunk 8 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java]: I also considered _dynamicCapacityMetricsMap / _mutableCapacityMetricsMap. I will change the name and also add more comments about the mutable map.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 13 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Discussed offline. Will consider making it simpler and removing the lock.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 16]
```diff
@@ -159,28 +163,30 @@ public synchronized void unregister() {
...
+      throws AttributeNotFoundException {
     if (SENSOR_NAME_TAG.equals(attribute)) {
       return getSensorName();
     }
 
-    if (!_attributeMap.containsKey(attribute)) {
-      return null;
+    DynamicMetric metric = _attributeMap.get(attribute);
+    if (metric == null) {
+      throw new AttributeNotFoundException("Attribute[" + attribute + "] is not found.");
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Have you confirmed what will be the output when the get method returns this Exception?
Here's what I want to test:
1. Make the MBeanInfo async with the list. See what is the value and if the fetching value from the MBean server works or not.
2. Fix the async part, then try to fetch the value again. See if the reporting logic still works fine and if we can fetch the right data again.
I just want to be safe.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 9 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Right, already added back and implemented.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java [hunk 17]
```diff
@@ -103,25 +100,30 @@ protected synchronized boolean doRegister(Collection<DynamicMetric<?, ?>> dynami
    * @param description description of the MBean
    * @param dynamicMetrics the DynamicMetrics
    */
-  private void updateAttributtInfos(Collection<DynamicMetric<?, ?>> dynamicMetrics,
+  protected void updateAttributesInfo(Collection<DynamicMetric<?, ?>> dynamicMetrics,
       String description) {
-    _attributeMap.clear();
+    if (dynamicMetrics == null || dynamicMetrics.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: I think so, as well. The original code checks null, so I do it to avoid changing the logic. If you also agree on not checking null, I can remove the check.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 16 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Let me do test it.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Check for null looks good to me. I mean "dynamicMetrics.isEmpty()" is not necessary.
This is a way to remove the metrics in the list.

[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Changed.

[the remark this answers is not in the document]
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Using an empty collection to remove the metrics is kind of hacky way to me. Let's confirm:
1. Do we allow null or empty to delete metrics? The original code allows this.
2. If we don't allow null to delete metrics, we will do nothing if it is null. We allow empty collection to delete metrics.

[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Or you can disallow it and add another method to clear the list. Both work for me : )

**Comment by jiajunwang** [comment]
@pkuwm So this PR has been pending for a while. When you have time, could you please rebase and notify us if it is ready again for the reviewers.

File: helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java (hunk 16 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/monitoring/mbeans/dynamicMBeans/DynamicMBeanProvider.java]: Test results:
1. MBeanInfo async with the list.
Running 2 threads: thread 1 awaits while it is updating attributes "A" and "B" (not completed); meanwhile, thread 2 updates attributes "C", "D" and completes. Finally, "C", "D" could be fetched correctly, while getting "A" and "B" throws AttributeNotFoundException. thread 1 completes updating. Fetching "A", "B" correctly, and throwing AttributeNotFoundException for fetching "C", "D"
2. Fix async.
Has run a few tests. It works fine and confirms to fetch the data correctly.

**Comment by pkuwm** [comment]
This PR is ready to be merged, approved by @jiajunwang 

```
We need to monitor instance utilization in purpose of understanding what the instance capacity is.

Change list:
- Change instance monitor to update capacity
- Change getAttribute to throw AttributeNotFoundException in DynamicMBeanProvider
- Combine max usage and instance capacity update into one method in cluster status monitor
- Add unit test
```
